graphql_api/types/test_analytics/test_analytics.py

```diff
@@ -1,24 +1,318 @@
 
+from codecov.commands.exceptions import ValidationError
 from codecov.db import sync_to_async
-from core.models import Repository
-from graphql_api.types.enums import OrderingDirection, TestResultsFilterParameter
+from graphql_api.types.enums import (
+    OrderingDirection,
+    TestResultsFilterParameter,
+    TestResultsOrderingParameter,
+)
 from graphql_api.types.enums.enum_types import MeasurementInterval
-from utils.test_results import (
-    GENERATE_TEST_RESULT_PARAM,
+from graphql_api.types.flake_aggregates.flake_aggregates import (
+    FlakeAggregates,
     generate_flake_aggregates,
-    generate_test_results,
+)
+from graphql_api.types.test_results_aggregates.test_results_aggregates import (
+    TestResultsAggregates,
     generate_test_results_aggregates,
-    get_flags,
-    get_test_suites,
 )
+from utils.test_results import get_results
 
 log = logging.getLogger(__name__)
 
+
+@dataclass
+class TestResultsRow:
+    # the order here must match the order of the fields in the query
+    name: str
+    test_id: str
+    testsuite: str | None
+    flags: list[str]
+    failure_rate: float
+    flake_rate: float
+    updated_at: dt.datetime
+    avg_duration: float
+    total_fail_count: int
+    total_flaky_fail_count: int
+    total_pass_count: int
+    total_skip_count: int
+    commits_where_fail: int
+    last_duration: float
+
+
+@dataclass
+class TestResultConnection:
+    edges: list[dict[str, str | TestResultsRow]]
+    page_info: dict
+    total_count: int
+
+
+DELIMITER = "|"
+
+
+@dataclass
+class CursorValue:
+    ordered_value: float | int | dt.datetime | str
+    name: str
+
+
+def decode_cursor(
+    value: str | None, ordering: TestResultsOrderingParameter
+) -> CursorValue | None:
+    if value is None:
+        return None
+
+    split_cursor = b64decode(value.encode("ascii")).decode("utf-8").split(DELIMITER)
+    ordered_value: str = split_cursor[0]
+    name: str = split_cursor[1]
+    match ordering:
+        case (
+            TestResultsOrderingParameter.AVG_DURATION
+            | TestResultsOrderingParameter.FLAKE_RATE
+            | TestResultsOrderingParameter.FAILURE_RATE
+            | TestResultsOrderingParameter.LAST_DURATION
+        ):
+            return CursorValue(ordered_value=float(ordered_value), name=name)
+        case TestResultsOrderingParameter.COMMITS_WHERE_FAIL:
+            return CursorValue(ordered_value=int(ordered_value), name=name)
+        case TestResultsOrderingParameter.UPDATED_AT:
+            return CursorValue(
+                ordered_value=dt.datetime.fromisoformat(ordered_value), name=name
+            )
+
+    raise ValueError(f"Invalid ordering field: {ordering}")
+
+
+def encode_cursor(row: TestResultsRow, ordering: TestResultsOrderingParameter) -> str:
+    return b64encode(
+        DELIMITER.join([str(getattr(row, ordering.value)), str(row.name)]).encode(
+            "utf-8"
+        )
+    ).decode("ascii")
+
+
+def validate(
```

Comment:
Random thought here, does the order of these validations matter? Do we want it to matter?

Reply:
nope, and i don't think it should matter (maybe in past iterations it would've(